[DataGrid] Add props to control cell mode (#4210)
mui/mui-x: docs/data/data-grid/editing/editing.md

The following demo shows how to prevent the user from exiting edit mode when clicking outside of a cell.
To do this, the `onCellEditStop` prop is used to check if the `reason` is `'cellFocusOut'`.
If that condition is true, it [disables](/x/react-data-grid/events/#disabling-the-default-behavior) the default event behavior.
In this context, the user can only stop editing a cell by pressing <kbd class="key">Enter</kbd>, <kbd class="key">Escape</kbd> or <kbd class="key">Tab</kbd>.

{{"demo": "DisableStopEditModeOnFocusOut.js", "bg": "inline"}}

## Controlled mode

Each cell and row has two modes: `edit` and `view`.
You can control the active mode using the props `cellModesModel` and `rowModesModel` (only works if `editMode="row"`).

The `cellModesModel` prop accepts an object containing the `mode` (and additional options) for a given column field, in a given row, as in the following example.

> The options accepted are the same available in `apiRef.current.startCellEditMode` and `apiRef.current.stopCellEditMode`.

Reviewer: ```suggestion
The options accepted are the same available in `[apiRef.current.startCellEditMode](#start-editing)` and `[apiRef.current.stopCellEditMode](#stop-editing)`.
```

```tsx
// Changes the mode of field=name from row with id=1 to "edit"
<DataGrid
  cellModesModel={{ 1: { name: { mode: GridCellModes.Edit } } }}
/>

// Changes the mode of field=name from row with id=1 to "view", ignoring modifications made
<DataGrid
  cellModesModel={{ 1: { name: { mode: GridCellModes.View, ignoreModifications: true } } }}
/>
```

For row editing, the `rowModesModel` props work in a similar manner.

> The options accepted are the same available in `apiRef.current.startRowEditMode` and `apiRef.current.stopRowEditMode`.

Reviewer: ```suggestion
The options accepted are the same available in `[apiRef.current.startRowEditMode](#start-editing)` and `[apiRef.current.stopRowEditMode](#stop-editing)`.
```

The following demo shows how to prevent the user from exiting edit mode when clicking outside of a cell.
To do this, the `onCellEditStop` prop is used to check if the `reason` is `'cellFocusOut'`.
If that condition is true, it [disables](/x/react-data-grid/events/#disabling-the-default-behavior) the default event behavior.
In this context, the user can only stop editing a cell by pressing <kbd class="key">Enter</kbd>, <kbd class="key">Escape</kbd> or <kbd class="key">Tab</kbd>.

{{"demo": "DisableStopEditModeOnFocusOut.js", "bg": "inline"}}

## Controlled mode

Each cell and row has two modes: `edit` and `view`.
You can control the active mode using the props `cellModesModel` and `rowModesModel` (only works if `editMode="row"`).

> The `cellModesModel` prop accepts an object where the key corresponds to the row ID containing the cell whose mode will be changed.

Author: The problem is that this interface suggested is not TS and the code will not be properly formatted. The correct interface would be 

```ts
export type GridCellModesModelProps =
  | ({ mode: GridCellModes.View } & Omit<GridStopCellEditModeParams, 'id' | 'field'>)
  | ({ mode: GridCellModes.Edit } & Omit<GridStartCellEditModeParams, 'id' | 'field'>);

export type GridCellModesModel = Record<GridRowId, Record<string, GridCellModesModelProps>>;

export type GridRowModesModelProps =
  | ({ mode: GridRowModes.View } & Omit<GridStopRowEditModeParams, 'id' | 'field'>)
  | ({ mode: GridRowModes.Edit } & Omit<GridStartRowEditModeParams, 'id' | 'field'>);

export type GridRowModesModel = Record<GridRowId, GridRowModesModelProps>;
```

but it doesn't make any sense.

> The `cellModesModel` prop accepts an object where the key corresponds to the row ID containing the cell whose mode will be changed.

Reviewer: Yeah. that interface is not really helpful in explaining it :).

Can we use jsx on this case to display the "template" object?

I think the paragraph describing the object is a bit convoluted. I tried simplifying it, but I don't think it's enough because the object itself is hard to explain while keeping the description accurate.

```suggestion
The `cellModesModel` prop accepts an object containing the `mode` (and additional options) for a given column field, in a given row, as in the example below.
The options accepted are the same available in apiRef.current.startCellEditMode and apiRef.current.stopCellEditMode.
```

> The `cellModesModel` prop accepts an object where the key corresponds to the row ID containing the cell whose mode will be changed.

Author: I applied the suggestion. We can only use JSX on JavaScript code. I think the example with the comments explain well the shape of the object.

Since cell editing works on cells, the value for the key is another object where the key is the column field of the cell.
Finally, the value for the column field key is another object containing `mode` and additional options.
The options accepted are the same available in `apiRef.current.startCellEditMode` and `apiRef.current.stopCellEditMode`.

```tsx
// Changes the mode of field=name from row with id=1 to "edit"
<DataGrid

>   cellModesModel={{ 1: { name: { mode: GridCellModes.Edit } } }}

Author: It doesn't look good because of https://github.com/PrismJS/prism/issues/1548

Reviewer 2: Agree!
